File: src/data/codes/multi-layer-perceptron.m
clc;  % multi-layer perceptron (patternnet)
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'mlp_disease_risk.csv';

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(24);
    n = 180;

    Age_Group = randi([0 2], n, 1);         % Young=0, Middle=1, Senior=2
    BMI_Category = randi([0 2], n, 1);      % Normal=0, Overweight=1, Obese=2
    Physical_Activity = randi([0 2], n, 1); % Low=0, Moderate=1, High=2
    Blood_Pressure = randi([0 2], n, 1);    % Normal=0, Elevated=1, High=2
    Family_History = randi([0 1], n, 1);    % No=0, Yes=1

    riskScore = 0.7*Age_Group + 1.0*BMI_Category - 0.8*Physical_Activity ...
                + 0.9*Blood_Pressure + 1.2*Family_History;
    Risk_Level = double(riskScore >= 2.5);

    data = table(Age_Group, BMI_Category, Physical_Activity, ...
                 Blood_Pressure, Family_History, Risk_Level);
else
    data = readtable(csvFile);
end

disp(head(data));

%% PREPROCESSING
data = rmmissing(data);
data = unique(data);

X = [data.Age_Group, data.BMI_Category, data.Physical_Activity, ...
     data.Blood_Pressure, data.Family_History]';
Y = data.Risk_Level';

%% TRAIN-TEST SPLIT
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(:, training(cv));
Y_train = Y(:, training(cv));
X_test  = X(:, test(cv));
Y_test  = Y(:, test(cv));

%% MODEL BUILDING
net = patternnet(10);   % one hidden layer with 10 neurons
net = train(net, X_train, Y_train);

%% PREDICTION
Y_prob = net(X_test);
Y_pred = round(Y_prob);

%% EVALUATION
confMat = confusionmat(Y_test', Y_pred');
accuracy = sum(diag(confMat)) / sum(confMat(:));
TP = confMat(2,2); FP = confMat(1,2); FN = confMat(2,1);
precision = TP/(TP+FP+eps);
recall = TP/(TP+FN+eps);
f1 = 2*precision*recall/(precision+recall+eps);

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nAccuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

figure;
confusionchart(Y_test', Y_pred');
title('MLP Confusion Matrix');

File: src/data/codes/multi-layer-perceptron.py
"""Multi-layer perceptron
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
RNG = np.random.default_rng(24)

if USE_SYNTHETIC:
    n = 180
    age = RNG.integers(0, 3, size=n)
    bmi = RNG.integers(0, 3, size=n)
    phys = RNG.integers(0, 3, size=n)
    bp = RNG.integers(0, 3, size=n)
    fam = RNG.integers(0, 2, size=n)
    risk = (
        0.7 * age
        + 1.0 * bmi
        - 0.8 * phys
        + 0.9 * bp
        + 1.2 * fam
    )
    risk_level = (risk >= 2.5).astype(int)
    data = pd.DataFrame(
        {
            "Age_Group": age,
            "BMI_Category": bmi,
            "Physical_Activity": phys,
            "Blood_Pressure": bp,
            "Family_History": fam,
            "Risk_Level": risk_level,
        }
    )
else:
    data = pd.read_csv("mlp_disease_risk.csv")

print(data.head())
data = data.dropna().drop_duplicates()
cols = ["Age_Group", "BMI_Category", "Physical_Activity", "Blood_Pressure", "Family_History"]
X = data[cols].values
y = data["Risk_Level"].values

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = MLPClassifier(
    hidden_layer_sizes=(10,),
    max_iter=1000,
    random_state=42,
)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print(f"Accuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

fig, ax = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax)
ax.set_title("MLP Confusion Matrix")
plt.show()
